web/packages/teleport/src/Discover/Server/DiscoveryConfigSsm/DiscoveryConfigSsm.tsx

```diff
@@ -0,0 +1,372 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState } from 'react';
+import {
+  Box,
+  Link as ExternalLink,
+  Text,
+  Flex,
+  ButtonSecondary,
+  Link,
+} from 'design';
+import styled from 'styled-components';
+import { Danger, Info } from 'design/Alert';
+import TextEditor from 'shared/components/TextEditor';
+import { ToolTipInfo } from 'shared/components/ToolTip';
+import FieldInput from 'shared/components/FieldInput';
+import { Rule } from 'shared/components/Validation/rules';
+import Validation, { Validator } from 'shared/components/Validation';
+import { makeEmptyAttempt, useAsync } from 'shared/hooks/useAsync';
+import { TextSelectCopyMulti } from 'shared/components/TextSelectCopy';
+
+import cfg from 'teleport/config';
+import { useDiscover } from 'teleport/Discover/useDiscover';
+import { Regions } from 'teleport/services/integrations';
+import { AwsRegionSelector } from 'teleport/Discover/Shared/AwsRegionSelector';
+import JoinTokenService, { JoinToken } from 'teleport/services/joinToken';
+
+import {
+  DISCOVERY_GROUP_CLOUD,
+  createDiscoveryConfig,
+  InstallParamEnrollMode,
+} from 'teleport/services/discovery';
+import { splitAwsIamArn } from 'teleport/services/integrations/aws';
+import useStickyClusterId from 'teleport/useStickyClusterId';
+
+import { ActionButtons, Header, Mark, StyledBox } from '../../Shared';
+
+import { DiscoveryConfigCreatedDialog } from './DiscoveryConfigCreatedDialog';
+
+const IAM_POLICY_NAME = 'EC2DiscoverWithSSM';
+
+export function DiscoveryConfigSsm() {
+  const { agentMeta, emitErrorEvent, nextStep, updateAgentMeta } =
+    useDiscover();
+
+  const { arnResourceName, awsAccountId } = splitAwsIamArn(
+    agentMeta.awsIntegration.spec.roleArn
+  );
+
+  const { clusterId } = useStickyClusterId();
+
+  const [selectedRegion, setSelectedRegion] = useState<Regions>();
+  const [ssmDocumentName, setSsmDocumentName] = useState(
+    'TeleportDiscoveryInstaller'
+  );
+  const [scriptUrl, setScriptUrl] = useState('');
+  const [createdToken, setCreatedToken] = useState<JoinToken>();
```

Comment:
If it's not used for rendering but merely in event handlers, then it could be a ref. This saves an unnecessary re-render and paints a better picture as to how this piece of state is used.

This also simplifies the code a tiny tiny bit, because someone reading the code doesn't have to understand how `joinToken` differs from `createdToken`. With a ref there can be a single `joinTokenRef`.

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleport/src/Discover/Server/DiscoveryConfigSsm/DiscoveryConfigSsm.tsx b/web/packages/teleport/src/Discover/Server/DiscoveryConfigSsm/DiscoveryConfigSsm.tsx
index 32c34ddefa..c622a6607d 100644
--- a/web/packages/teleport/src/Discover/Server/DiscoveryConfigSsm/DiscoveryConfigSsm.tsx
+++ b/web/packages/teleport/src/Discover/Server/DiscoveryConfigSsm/DiscoveryConfigSsm.tsx
@@ -16,7 +16,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import React, { useState } from 'react';
+import React, { useState, useRef } from 'react';
 import {
   Box,
   Link as ExternalLink,
@@ -70,7 +70,7 @@ export function DiscoveryConfigSsm() {
     'TeleportDiscoveryInstaller'
   );
   const [scriptUrl, setScriptUrl] = useState('');
-  const [createdToken, setCreatedToken] = useState<JoinToken>();
+  const joinTokenRef = useRef<JoinToken>();
   const [showRestOfSteps, setShowRestOfSteps] = useState(false);
 
   const [attempt, createJoinTokenAndDiscoveryConfig, setAttempt] = useAsync(
@@ -80,14 +80,12 @@ export function DiscoveryConfigSsm() {
         // Don't create another token if token was already created.
         // This can happen if creating discovery config attempt failed
         // and the user retries.
-        let joinToken = createdToken;
-        if (!joinToken) {
-          joinToken = await joinTokenService.fetchJoinToken({
+        if (!joinTokenRef.current) {
+          joinTokenRef.current = await joinTokenService.fetchJoinToken({
             roles: ['Node'],
             method: 'iam',
             rules: [{ awsAccountId }],
           });
-          setCreatedToken(joinToken);
         }
 
         const config = await createDiscoveryConfig(clusterId, {
@@ -105,7 +103,7 @@ export function DiscoveryConfigSsm() {
               install: {
                 enrollMode: InstallParamEnrollMode.Script,
                 installTeleport: true,
-                joinToken: joinToken.id,
+                joinToken: joinTokenRef.current.id,
               },
             },
           ],
@@ -138,7 +136,7 @@ export function DiscoveryConfigSsm() {
 
   function clear() {
     setAttempt(makeEmptyAttempt);
-    setCreatedToken(null);
+    joinTokenRef.current = undefined;
   }
 
   return (

```
</details>

Reply:
this is nice, ty 👍